examples/simple-inference/simple-inference.cpp

```diff
@@ -0,0 +1,572 @@
+        }
+        fflush(stdout);
+
+        tokens_curr += batch.n_tokens;
+    }
+    console::set_display(console::reset);
+    return true;
+}
+
+
+static void dump_batches(
+        const llama_context *ctx,
+        const std::vector<struct batch_state> & batch_states,
+        const size_t prompt_size,
+        const size_t prompt_start = 0) {
+
+    for (size_t i = batch_states.size(); i > 0; i--) {
+        const batch_state & bs = batch_states[i - 1];
+
+        console::set_display(console::reset);
+        printf("\n\n%s Result #%d (size: %zu",
+            i != batch_states.size() ? "====================" : "####################",
+            bs.seq_id + 1, bs.last_tokens.size());
+#ifndef LLAMA_NO_SEQREP_SAMPLER
+        printf(", rewind cnt/toks: %zu/%zu", bs.rewind_count, bs.rewind_tokens);
+#endif
+        puts("):");
+        console::set_display(console::prompt);
+
+        for (size_t ti = prompt_start; ti < bs.last_tokens.size(); ti++) {
+            const std::string token_str = llama_token_to_piece(ctx, bs.last_tokens[ti]);
+            if (ti == prompt_size) console::set_display(console::reset);
+            fputs(token_str.c_str(), stdout);
+        }
+        fflush(stdout);
+    }
+}
+
+int main(int argc, char ** argv) {
+    gpt_params params;
+
+    if (gpt_params_parse(argc, argv, params) == false) {
+        return 1;
+    }
+
+    if (!check_unsupported(&params)) {
+        return 1;
+    }
+
+#ifndef LOG_DISABLE_LOGS
+    log_set_target(log_filename_generator("simple-inference", "log"));
+    LOG_TEE("Log start\n");
+    log_dump_cmdline(argc,argv);
+#endif // LOG_DISABLE_LOGS
+
+    llama_context *     ctx = NULL;
+    llama_model *     model = NULL;
+    llama_grammar * original_grammar = NULL;
+    std::vector<llama_token> prompt_tokens;
+
+    if (!initialize(&ctx, &model, params, prompt_tokens, &original_grammar)) {
+        return 1;
+    }
+
+    const int n_ctx   = llama_n_ctx(ctx);
+    const int n_vocab = llama_n_vocab(llama_get_model(ctx));
+    int n_remain      = params.n_predict;
+
+    const size_t prompt_size = prompt_tokens.size();
+    llama_batch batch = llama_batch_init(std::max(int32_t(prompt_size), params.n_batch), 0);
+
+    if (!feed_prompt(ctx, params, batch, prompt_tokens)) {
+        return 1;
+    }
+
+    if (n_remain < 0 || n_remain + int(prompt_size) > n_ctx) {
+        n_remain = n_ctx - prompt_size;
+    }
+    std::vector<struct batch_state> batch_states;
+    batch_states.reserve(params.n_parallel);
+    for (int32_t i = 0; i < params.n_parallel; i++) {
+        struct batch_state bs = {
+            llama_seq_id(i),
+            batch.n_tokens - 1,
+            size_t(n_remain),
+            prompt_tokens,
+            original_grammar != NULL ? llama_grammar_copy(original_grammar) : NULL,
+#ifndef LLAMA_NO_SEQREP_SAMPLER
+            size_t(prompt_size),
+            seqrep_rewind_state(n_vocab, n_ctx, 2000, prompt_size - 1),
+            size_t(0),
+            size_t(0),
```

Comment:
These size_t casts are not needed. The one above is because n_remain is an int.

Reply:
Thanks for the feedback. Honestly, I am not really making much of an attempt to write nice code at this point. Once I actually know what features I'm keeping then I intended to perform a serious cleanup pass.

I am still really interested in hearing about any actual bugs, like off by one errors, etc.

By the way, spare your virgin eyes from looking at the version I just pushed. You might just go blind if you don't heed the warning. It's bad, really bad.